pkg/ccl/logictestccl/tests/local-vec-off/local-vec-off_test: TestCCLLogic_select_for_update_read_committed failed

pkg/ccl/logictestccl/tests/local-vec-off/local-vec-off_test.TestCCLLogic_select_for_update_read_committed [failed](https://tanzanite.cluster.engflow.com/invocations/default/74576e32-820b-47a6-841f-ac24652f0abb?testReportRun=14&testReportShard=22&testReportAttempt=1#targets-Ly9wa2cvY2NsL2xvZ2ljdGVzdGNjbC90ZXN0cy9sb2NhbC12ZWMtb2ZmOmxvY2FsLXZlYy1vZmZfdGVzdA==) on master @ [78fa126dfc749e620c28531c1659e35bf98034ec](https://github.com/cockroachdb/cockroach/commits/78fa126dfc749e620c28531c1659e35bf98034ec):


```
=== RUN   TestCCLLogic_select_for_update_read_committed
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestCCLLogic_select_for_update_read_committed3254322103
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:144: cluster virtualization disabled in global scope due to issue: #76378 (expected label: C-bug)
[10:21:36] setting distsql_workmem='15614B';
    logic.go:4317: -- test log scope end --
test logs left over in: outputs.zip/logTestCCLLogic_select_for_update_read_committed3254322103
--- FAIL: TestCCLLogic_select_for_update_read_committed (1.08s)
=== RUN   TestCCLLogic_select_for_update_read_committed/skip_locked
    logic.go:2967: 
         
        /var/lib/engflow/worker/work/3/exec/bazel-out/k8-fastbuild/bin/pkg/ccl/logictestccl/tests/local-vec-off/local-vec-off_test_/local-vec-off_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/logictestccl/testdata/logic_test/select_for_update_read_committed:607: SELECT * FROM xyz WHERE z = 100 ORDER BY x FOR UPDATE SKIP LOCKED
        expected:
            2  10  100
            
        but found (query options: "") :
            
    logic.go:2205: 
         /var/lib/engflow/worker/work/3/exec/bazel-out/k8-fastbuild/bin/pkg/ccl/logictestccl/tests/local-vec-off/local-vec-off_test_/local-vec-off_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/logictestccl/testdata/logic_test/select_for_update_read_committed:612: too many errors encountered, skipping the rest of the input
[10:21:36] --- done: /var/lib/engflow/worker/work/3/exec/bazel-out/k8-fastbuild/bin/pkg/ccl/logictestccl/tests/local-vec-off/local-vec-off_test_/local-vec-off_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/logictestccl/testdata/logic_test/select_for_update_read_committed with config local-vec-off: 106 tests, 2 failures
--- total: 106 tests, 2 failures
    --- FAIL: TestCCLLogic_select_for_update_read_committed/skip_locked (0.02s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=14</code>
 - <code>shard=22</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCCLLogic_select_for_update_read_committed.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-40849

--- a/pkg/ccl/logictestccl/testdata/logic_test/select_for_update_read_committed
+++ b/pkg/ccl/logictestccl/testdata/logic_test/select_for_update_read_committed
@@ -576,6 +576,9 @@ INSERT INTO xyz VALUES (1, 10, 100), (2, 10, 100)
 user testuser
 
 statement ok
+SET optimizer_use_lock_op_for_serializable = true
+
+statement ok
 BEGIN ISOLATION LEVEL SERIALIZABLE
 
 # Lock the first row.
@@ -613,4 +616,10 @@ user testuser
 statement ok
 COMMIT
 
+statement ok
+RESET optimizer_use_lock_op_for_serializable
+
 user root
+
+statement ok
+RESET optimizer_use_lock_op_for_serializable

--- a/pkg/sql/exec_factory_util.go
+++ b/pkg/sql/exec_factory_util.go
@@ -273,7 +273,7 @@ func constructVirtualScan(
 	if params.NeededCols.Contains(0) {
 		return nil, errors.Errorf("use of %s column not allowed.", table.Column(0).ColName())
 	}
-	if params.Locking.IsNonZeroLocking() {
+	if !params.Locking.IsNoOp() {
 		// We shouldn't have allowed SELECT FOR UPDATE for a virtual table.
 		return nil, errors.AssertionFailedf("locking cannot be used with virtual table")
 	}

--- a/pkg/sql/opt/exec/execbuilder/mutation.go
+++ b/pkg/sql/opt/exec/execbuilder/mutation.go
@@ -1227,7 +1227,7 @@ func (b *Builder) buildLock(lock *memo.LockExpr) (_ execPlan, outputCols colOrdM
 	if err != nil {
 		return execPlan{}, colOrdMap{}, err
 	}
-	if !locking.IsNonZeroLocking() {
+	if locking.IsNoOp() {
 		return b.buildRelational(lock.Input)
 	}
 

--- a/pkg/sql/opt/locking.go
+++ b/pkg/sql/opt/locking.go
@@ -82,17 +82,18 @@ func (l Locking) Max(l2 Locking) Locking {
 	}
 }
 
-// IsLocking returns whether the receiver is configured to use a row-level
-// locking mode.
+// IsLocking returns whether the receiver is configured to use row-level
+// locking.
 func (l Locking) IsLocking() bool {
 	return l.Strength != tree.ForNone
 }
 
-// IsNonZeroLocking returns true if the locking properties somehow differ from
-// the zero value locking properties. This can mean a row-level locking mode is
-// set (i.e. l.Strength != tree.ForNone) or the SKIP LOCKED wait policy is in
-// use even if no locking mode is set (i.e. l.WaitPolicy ==
-// tree.LockWaitSkipLocked).
-func (l Locking) IsNonZeroLocking() bool {
-	return l != Locking{}
+// IsNoOp returns true if none of the locking properties are set. It differs
+// from IsLocking in that it considers all of the locking properties, instead of
+// only Strength. Currently, the only locking property that can be set when
+// Strength=ForNone is WaitPolicy=LockWaitSkipLocked. So we can say: IsNoOp
+// returns false if IsLocking returns true OR the SKIP LOCKED wait policy is in
+// effect.
+func (l Locking) IsNoOp() bool {
+	return l == Locking{}
 }

--- a/pkg/sql/opt/memo/expr_format.go
+++ b/pkg/sql/opt/memo/expr_format.go
@@ -1629,7 +1629,7 @@ func (f *ExprFmtCtx) formatLocking(tp treeprinter.Node, locking opt.Locking) {
 func (f *ExprFmtCtx) formatLockingWithPrefix(
 	tp treeprinter.Node, labelPrefix string, locking opt.Locking,
 ) {
-	if !locking.IsNonZeroLocking() {
+	if locking.IsNoOp() {
 		return
 	}
 	strength := ""

--- a/pkg/sql/opt/memo/logical_props_builder.go
+++ b/pkg/sql/opt/memo/logical_props_builder.go
@@ -81,7 +81,7 @@ func (b *logicalPropsBuilder) buildScanProps(scan *ScanExpr, rel *props.Relation
 	// Side Effects
 	// ------------
 	// A Locking option is a side-effect (we don't want to elide this scan).
-	if scan.Locking.IsNonZeroLocking() {
+	if !scan.Locking.IsNoOp() {
 		rel.VolatilitySet.AddVolatile()
 	}
 

